src/lib/SaferNames/NameCore.hs

```diff
@@ -0,0 +1,294 @@
+-- a list, because that's all we need at runtime. But it's important to remember
+-- that it conceptually represents a list. For example, a `NameSet n` and a `NameSet
+-- m` that happen to represent the same set of names can't necessarily be
+-- considered equal.) Types of kind `S` are mostly created existentially through
+-- rank-2 polymorphism, rather than using the constructors in the data
+-- definition. For example: magicallyCreateFreshS :: (forall (n::S). a) -> a
+-- magicallyCreateFreshS x = x -- where does `n` come from? magic!
+
+-- We also have `:=>:` to represent differences between scopes with a common
+-- prefix. A `NameSet (n:=>:l)` means that
+--   1. `NameSet n` is a prefix of `NameSet l`
+--   2. `NameSet (n:=>:l)` is the list of names by which `l` extends `n`.
+
+--      x    y    z    x    w    x
+--     \-----------------/\--------/
+--              n           n:=>:l
+--     \---------------------------/
+--                    l
+
+-- Note that `l` is not necessarily a *fresh* extension: in the example above, x
+-- appears in `n:=>:l` even though it already appeared, twice, in `n`.
+-- We have separate proof objects, `Ext n l` and `Distinct n` to provide
+-- evidence of freshness.
+
+-- There are also special scopes, `VoidS` and `UnitS`, representing the
+-- empty list and a singleton list with a particular special name. These are
+-- useful in the same way that the ordinary types `Void` and `()` are useful.
+
+data S = (:=>:) S S
+       | VoidS
+       | UnsafeMakeS
+       | UnsafeMakeDistinctS
+
+type E = S -> *       -- expression-y things, covariant in the S param
+type B = S -> S -> *  -- binder-y things, covariant in the first param and
+                      -- contravariant in the second. These are things like
+                      -- `Binder n l` or `Decl n l`, that bind the names in
+                      -- `NameSet (n:=>:l)`, extending `n` to `l`. Their free
+                      -- name are in `NameSet n`. We sometimes call `n` the
+                      -- "outside scope" and "l" the "inside scope".
+
+newtype NameSet (n::S) = UnsafeMakeNameSet (S.Set RawName)
+
+emptyNameSetFrag :: NameSet (n:=>:n)
+emptyNameSetFrag = UnsafeMakeNameSet mempty
+
+emptyNameSet :: NameSet VoidS
+emptyNameSet = UnsafeMakeNameSet mempty
+
+extendNameSet :: NameSet n -> NameSet (n:=>:l) -> NameSet l
+extendNameSet (UnsafeMakeNameSet s1) (UnsafeMakeNameSet s2) =
+  UnsafeMakeNameSet (s1 <> s2)
+
+concatNameSets :: NameSet (n1:=>:n2) -> NameSet (n2:=>:n3) -> NameSet (n1:=>:n3)
+concatNameSets (UnsafeMakeNameSet s1) (UnsafeMakeNameSet s2) =
+  UnsafeMakeNameSet (s1 <> s2)
+
+-- TODO: we reuse the old `Name` to make use of the GlobalName name space while
+-- we're using both the old and new systems together.
+-- TODO: something like this instead:
+--    type Tag = T.Text
+--    data RawName = RawName Tag Int deriving (Show, Eq, Ord)
+type RawName = D.Name
+
+data Name (s::E)  -- static information associated with name
```

Comment:
Yes, static information is what we currently call `BinderInfo`. It includes type info , but also things like the flavor of arrow if it's a lambda-bound variable, or the actual rhs of the let binding if it's let-bound. `s` has kind `S -> *` because it's parameterized by a name space. For example, if the static information is a type, then it may contain free variables.

Reply:
Ah, that makes sense. Might be helpful to include that explanation as a comment?